### DISCUSSION TITLE
Luhmann-like file-naming scheme

### DISCUSSION THREAD

**Discussion by gour** [getreu/tp-note#16] [Q&A] [answered]
What do you think about this: recently I watched the video [series](https://www.thesing-online.de/blog/denote-as-a-zettelkasten) which describes Zettelkasten note-taking by using Emacs' [Denote](https://protesilaos.com/emacs/denote) package. It uses a very interesting file-naming scheme in the signature branch allowing one to name one's notes with the “signatures” which do resemble, afaict, the original Luhmann's index-card's naming scheme. This allows to “…be used to establish a sequential relationship between files (e.g. 1, 1a, 1b, 1b1, 1b2, …)”, so I wonder if something like that could be deployed with Tp-note?

Denote package itself is very powerful, but I'm simply still not sold on Emacs…not knowing Elisp etc….

**Comment by gour** [comment]
I mostly wonder about the ability to use `id`s like "1, 1a, 1b..." in the note's template header or as a part of the filename for easy sorting, which does resemble Luhman's index card naming  scheme?

**Comment by getreu** [comment]
Dear Saša,

I had a look into the videos you sent me, thank you!

> I mostly wonder about the ability to use IDs like "1, 1a, 1b..." in
> the note's template header or as a part of the filename for easy
> sorting, which does resemble Luhman's index card naming scheme?

Yes, you can. To allow alphanumerical IDs inside your sort tags
change Tp-Note's configuration file variable `filename.sort_tag_chars`
from `sort_tag_chars = '0123456789.-_ 	'`
to `sort_tag_chars = '0123456789abcdefghijklmnopqrstuvwxyz-='`
The above definition allows you to include Zettelkasten IDs into your sort
tags.

### The basic workflow is

1. Write your Tp-Note note as usual. Save it.
2. Change the sort tag (signature) part of the filename in your file manager
   and include the above IDs, e.g. change the filename `20230416-My Note.md` 
   to `20230416=1c2a-My Note.md` or, more short to `1c2a-My Note.md` if
   you prefer.

### Alternative workflow
1. Write your Tp-Note note as usual. Don't save it yet.
2. Add the header field `sort_tag: 1c2a-` to the YAML header
   of your note.
3. Save the file. Tp-Note will produce the file `1c2a-My Note.md`

### Remarks and links

1. In Tp-Note parlance the term "signature" is called "sort tag".
   More here: [sort tags]
2. Filename generation in Tp-Note is controlled by the so-called customizable
   _filename templates_. These are written in the Tera programming language.
   More here: [template types]
3. You can instruct Tp-Note to simulate Denote's filename scheme (when creating new
   notes), ~~by changing the filename templates in Tp-Note's configuration file~~ with the command
   line argument  `--scheme zettel`.
 
[sort tags]: https://blog.getreu.net/projects/tp-note/tpnote--manpage.html#metadata-filename-synchronization
[template types]: https://blog.getreu.net/projects/tp-note/tpnote--manpage.html#template-types


### Are Zettelkasten IDs still useful? 

As Stefan Thesing mentions in his first video, there is a controversy among 
Zettelkasten users, if Zettelkasten IDs still make sense using computers.
I do fine without them. I guess Zettelkasten purists certainly want them.

**Reply by gour** [reply]
Hello Jens,

> I had a look into the videos you sent me, thank you!

I'm glad you found them useful. :smile: 

> Yes, you can. To allow alphanumerical IDs inside your sort tags change Tp-Note's configuration file variable `filename.sort_tag_chars` from `sort_tag_chars = '0123456789.-_ '` to `sort_tag_chars = '0123456789abcdefghijklmnopqrstuvwxyz-='` The above definition allows you to include Zettelkasten IDs into your sort tags.

Hmm, that is great!

>     1. In Tp-Note parlance the term "signature" is called "sort tag".
>        More here: [sort tags](https://blog.getreu.net/projects/tp-note/tpnote--manpage.html#metadata-filename-synchronization)
> 
>     2. Filename generation in Tp-Note is controlled by the so-called customizable
>        _filename templates_. These are written in the Tera programming language.
>        More here: [template types](https://blog.getreu.net/projects/tp-note/tpnote--manpage.html#template-types)

I'll definitely try it out!

> ### Are Zettelkasten IDs still useful?
> 
> As Stefan Thesing mentions in his first video, there is a controversy among Zettelkasten users, if Zettelkasten IDs still make sense using computers. I do fine without them. I guess Zettelkasten purists certainly want them.

I believe they're since they connect the notes not by just dates or hyperlinks, but they provide a “context” how one has arrived to certain train of thought…

**Comment by getreu** [comment]
Hello Saša,

as I understand, Zettelkasten people are very rigorous about Note IDs. They also have the habit to extensively link notes to other notes. 

Tp-Note version 1.21.3 comes with a new feature simplifying document interlinking. Instead of giving the full file name of the link's destination `[my note](<1a-2a=My Note.md>)`, now the ID (sort-tag) of the destination file, e.g. `[my note](1a-2a)` suffices.
The same applies to autolinks: `<http:1a-2a=My Note.md>` can now be shortened to `<http:1a-2c>`.

**Comment by gour** [comment]
On Wed, May 10, 2023 at 04:28:20AM -0700, Jens Getreu wrote:

Hello Jens,

> as I understand, Zettelkasten people are very rigorous about Note IDs. 
> They also have the habit to extensively link notes to other notes.

You're right!

> Tp-Note version 1.21.3 comes with a new feature simplifying document 
> interlinking. Instead of giving the full file name of the link's 
> destination `[my note](<1a-2a=My Note.md>)`, now the ID (sort-tag) of 
> the destination file, e.g. `[my note](1a-2a)` suffices.
> The same applies to autolinks: `<http:1a-2a=My Note.md>` can now be 
> shortened to `<http:1a-2c>`.

This is very interesting...I'm still considering whether to settle on 
Emacs-based setup or to combine several apps into more vim-oriented one.


Sincerely,
Gour

-- 
He is a perfect yogī who, by comparison to his own self,
sees the true equality of all beings, in both their
happiness and their distress, O Arjuna!

**Comment by getreu** [comment, marked as answer]
Hello Saša @gour  

I wrote a article on how to set up (and use) a Zettelkasten with Tp-Note.
[Set up a Zettelkasten with Tp-Note](https://blog.getreu.net/20230904-tp-note-news11/)

**Comment by gour** [comment]
On Thu, 30 Nov 2023 12:17:39 -0800
Jens Getreu ***@***.***> wrote:

Hello Jens!

I apologize for late reply...after seeing your mail just skimmed over
the article without having time to go deeper into it..
> I wrote a article on how to set up (and use) a Zettelkasten with
> Tp-Note. [Set up a Zettelkasten with
> Tp-Note](https://blog.getreu.net/20230904-tp-note-news11/)

Yesterday I took time to scrutinize it!

After going slowly through it, I must admit that it helped me resolve
some issue - for some time I was considering whether to use Denote or
another Emacs package called ekg (https://github.com/ahyatt/ekg) which
is using sqlite3 database. For my other writing I've settled on using
Emacs and probably org-mode, but "The fact, that Tp-Note note files can
be opened with any text editor and without any additional proprietary
software, makes our Zettelkasten long-term future-proof." clinged in my
mind to decide in favour of more text-oriented Denote.

However, it's interesting that you're mentioning Denote several times
in your article and wonder if Tp-Note could be made more
markup-agnostic like Denote and/or to support e.g. org-mode header?

Maybe it's already possible by creating proper template - still have to
do practical session based on your article to research what are
possibilities to have interoperation between Denote and Tp-Note?


Sincerely,
Gour (wishing you happy holidays!)

-- 
Even if you are considered to be the most sinful of all sinners,
when you are situated in the boat of transcendental knowledge
you will be able to cross over the ocean of miseries.

**Comment by getreu** [comment]
Hello Saša,

> if Tp-Note could be made more markup-agnostic like Denote

Yes, it is (cf. [Change the default markup language](https://blog.getreu.net/projects/tp-note/tpnote--manpage.html#change-the-default-markup-language))

>  and/or to support e.g. org-mode header?

No, the header syntax must be [YAML](https://www.redhat.com/en/topics/automation/what-is-yaml), but you can define the names, meaning and localization of the header variables. Tp-Note v1.23 comes with two different predefined  schemes: `default` and `zettel`. The `zettel` scheme is compatible with Denote. You can add own schemes as long as the header is YAML.

Merry Christmas
